### PR TITLE
fix(app): add captive fatal error modal to LPC, cancel run on close 

### PR DESCRIPTION
**Pull request by b-cooper**
# Overview

If an error occurs during LPC, the safest thing to do is to report the message to the user and
cancel the run.

Closes RQA-592, Closes RQA-593, Close RQA-583

# Changelog

# Review requests

# Risk assessment
med

### CODE REVIEW

File: app/src/organisms/LabwarePositionCheck/CheckItem.tsx [hunk 1]
```diff
@@ -97,7 +99,9 @@ export const CheckItem = (props: CheckItemProps): JSX.Element | null => {
       chainRunCommands(modulePrepCommands, true)
         .then(() => {})
         .catch(e => {
-          console.error('Unexpected command failure: ', e)
+          setFatalError(
+            `CheckItem module prep commands failed with message: ${e}`
```
Comment (shlokamin) [inline app/src/organisms/LabwarePositionCheck/CheckItem.tsx]: i think the string interpolation you have here will work but safer to pull out the message from the error object: 
```suggestion
            `CheckItem module prep commands failed with message: ${e.message}`
```

File: app/src/organisms/LabwarePositionCheck/CheckItem.tsx [hunk 2]
```diff
@@ -180,11 +184,15 @@ export const CheckItem = (props: CheckItemProps): JSX.Element | null => {
             })
           })
           .catch(e => {
-            console.error('Unexpected command failure: ', e)
```
Comment (shlokamin) [inline app/src/organisms/LabwarePositionCheck/CheckItem.tsx:183]: same deal here

File: app/src/organisms/LabwarePositionCheck/CheckItem.tsx [hunk 3]
```diff
@@ -180,11 +184,15 @@ export const CheckItem = (props: CheckItemProps): JSX.Element | null => {
...
+            setFatalError(
+              `CheckItem failed to save position for initial placement with message: ${e}`
```
Comment (shlokamin) [inline app/src/organisms/LabwarePositionCheck/CheckItem.tsx]: and here

File: app/src/organisms/LabwarePositionCheck/CheckItem.tsx [hunk 4]
```diff
@@ -180,11 +184,15 @@ export const CheckItem = (props: CheckItemProps): JSX.Element | null => {
...
+            )
           })
       })
       .catch(e => {
-        console.error('Unexpected command failure: ', e)
+        setFatalError(
+          `CheckItem failed to move to initial position with message: ${e}`
```
Comment (shlokamin) [inline app/src/organisms/LabwarePositionCheck/CheckItem.tsx]: and here

File: app/src/organisms/LabwarePositionCheck/CheckItem.tsx [hunk 5]
```diff
@@ -237,11 +245,15 @@ export const CheckItem = (props: CheckItemProps): JSX.Element | null => {
         chainRunCommands(confirmPositionCommands, true)
           .then(() => proceed())
           .catch(e => {
-            console.error('Unexpected command failure: ', e)
+            setFatalError(
+              `CheckItem failed to move from final position with message: ${e}`
```
Comment (shlokamin) [inline app/src/organisms/LabwarePositionCheck/CheckItem.tsx]: and here

File: app/src/organisms/LabwarePositionCheck/CheckItem.tsx [hunk 6]
```diff
@@ -237,11 +245,15 @@ export const CheckItem = (props: CheckItemProps): JSX.Element | null => {
...
+            )
           })
       })
       .catch(e => {
-        console.error('Unexpected command failure: ', e)
+        setFatalError(
+          `CheckItem failed to save final position with message: ${e}`
```
Comment (shlokamin) [inline app/src/organisms/LabwarePositionCheck/CheckItem.tsx]: and here

File: app/src/organisms/LabwarePositionCheck/CheckItem.tsx [hunk 7]
```diff
@@ -258,7 +270,7 @@ export const CheckItem = (props: CheckItemProps): JSX.Element | null => {
         })
       })
       .catch(e => {
-        console.error('Unexpected command failure: ', e)
+        setFatalError(`CheckItem failed to home: ${e}`)
```
Comment (shlokamin) [inline app/src/organisms/LabwarePositionCheck/CheckItem.tsx]: and here

File: app/src/organisms/LabwarePositionCheck/IntroScreen/index.tsx [hunk 8]
```diff
@@ -27,16 +27,25 @@ export const IntroScreen = (props: {
...
+  } = props
   const { t } = useTranslation(['labware_position_check', 'shared'])
   const handleClickStartLPC = (): void => {
     const prepCommands = getPrepCommands(protocolData)
     chainRunCommands(prepCommands, true)
       .then(() => proceed())
       .catch(e => {
-        console.error('Unexpected command failure: ', e)
+        setFatalError(
+          `IntroScreen failed to issue prep commands with message: ${e}`
```
Comment (shlokamin) [inline app/src/organisms/LabwarePositionCheck/IntroScreen/index.tsx]: here too

File: app/src/organisms/LabwarePositionCheck/PickUpTip.tsx [hunk 9]
```diff
@@ -145,11 +147,15 @@ export const PickUpTip = (props: PickUpTipProps): JSX.Element | null => {
             })
           })
           .catch(e => {
-            console.error('Unexpected command failure: ', e)
+            setFatalError(
+              `PickUpTip failed to save position for initial placement with message: ${e}`
```
Comment (shlokamin) [inline app/src/organisms/LabwarePositionCheck/PickUpTip.tsx]: and here

File: app/src/organisms/LabwarePositionCheck/PickUpTip.tsx [hunk 10]
```diff
@@ -145,11 +147,15 @@ export const PickUpTip = (props: PickUpTipProps): JSX.Element | null => {
...
+            )
           })
       })
       .catch(e => {
-        console.error('Unexpected command failure: ', e)
+        setFatalError(
+          `PickUpTip failed to move to initial position with message: ${e}`
```
Comment (shlokamin) [inline app/src/organisms/LabwarePositionCheck/PickUpTip.tsx]: and here

File: app/src/organisms/LabwarePositionCheck/PickUpTip.tsx [hunk 11]
```diff
@@ -184,11 +190,15 @@ export const PickUpTip = (props: PickUpTipProps): JSX.Element | null => {
...
+            setFatalError(
+              `PickUpTip failed to move from final position with message: ${e}`
+            )
           })
       })
       .catch(e => {
-        console.error('Unexpected command failure: ', e)
+        setFatalError(
+          `PickUpTip failed to save final position with message: ${e}`
+        )
```
Comment (shlokamin) [inline app/src/organisms/LabwarePositionCheck/PickUpTip.tsx]: and here

**Review by shlokamin — APPROVED** [review]
LGTM! for some of these errors it would be nice for the robot server to give us more info, but i guess this is all we have. way better than not doing anything though

Screen Shot 2023-03-28 at 5 44 42 PM

**Review by smb2268 — APPROVED** [review]
I wasn't able to test this on a robot but the code looks sound to me, `chainRunCommand` updates look good!

**Pull request merged**